### PR TITLE
feat: Refactor access control logic for all pages

### LINKED ISSUE #272: [FEATURE]: Members tab and table should be visible only to owners and maintainers
### Description

Currently, all users have access to members tab.

### Possible solution

- Render members table to Owners and maintainers only
- Render 404 to guests and developer
- Hide the top members tab to devs / guests

You can just return `notFound: true` on `getServerSideProps` to render 404. Here is an example

image

### Alternatives

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by ttebify**
# Description

> Please include a summary of the changes and the related issue. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes #272

## Type of change

> Please delete options that are not relevant.

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# How Has This Been Tested?

> Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

**Test Configuration**:

- OS with version:
- Browser with version:
- SDK / CLI with version:

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I documented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: apps/platform/components/members/Table.tsx [hunk 1]
```diff
@@ -2,6 +2,7 @@ import { useRouter } from "next/router";
 import { useCallback, useEffect, useState } from "react";
 import React from "react";
 import useFuse from "@/hooks/useFuse";
+import { useUserAccessStore } from "@/store/UserRole";
```
Comment (dahal) [inline apps/platform/components/members/Table.tsx:5]: I dont think this is being used right?

File: apps/platform/components/members/Table.tsx [hunk 2]
```diff
@@ -18,7 +19,7 @@ interface TableProps {
   tab: Tab;
   setTab: (tab: Tab) => void;
   members: UserType[];
-  userAccessInCurrentProject: Access;
+  roleInCurrentProject: UserRole;
```
Comment (dahal) [inline apps/platform/components/members/Table.tsx:22]:
```suggestion
  currentRole: UserRole;
```

Lets follow common pattern, eg: `currentProject` , `currentUser`, `currentRole` and so on.

File: apps/platform/utils/withAccessControl.ts [hunk 3]
```diff
@@ -4,28 +4,31 @@ import type {
...
-  projectRole: any;
+  currentProject: Project;
+  projects: Project[];
+  roleInProject: UserRole;
 };
 
 type AccessControlParams<P> = {
-  getServerSideProps?: GetServerSideProps<P & PageProps>;
-  checkProjectOwner?: boolean;
+  hasAccess?: Partial<Record<UserRole, boolean>>;
```
Comment (dahal) [inline apps/platform/utils/withAccessControl.ts:20]: Love it 😻

File: apps/platform/utils/withAccessControl.ts [hunk 4]
```diff
@@ -71,24 +74,30 @@ export function withAccessControl<P = Record<string, unknown>>({
...
-          },
-        };
+      return {
+        notFound: true,
+      };
+    }
+
+    let authorized = false;
+
+    Object.entries(hasAccess).forEach(([role, authorizedForRole]) => {
```
Comment (dahal) [inline apps/platform/utils/withAccessControl.ts:88]:
```suggestion
    Object.entries(hasAccess).forEach(([role, authorized]) => {
```

Comment (dahal) [inline apps/platform/utils/withAccessControl.ts:88]:
```suggestion
    Object.entries(hasAccess).forEach(([role, canAccess]) => {
```

File: apps/platform/utils/withAccessControl.ts [hunk 5]
```diff
@@ -71,24 +74,30 @@ export function withAccessControl<P = Record<string, unknown>>({
...
+      if (authorizedForRole && currentProject.role === role) {
```
Comment (dahal) [inline apps/platform/utils/withAccessControl.ts:89]:
```suggestion
      if (canAccess && currentProject.role === role) {
```

File: apps/platform/store/UserRole.ts [hunk 6]
```diff
@@ -0,0 +1,17 @@
+import { UserRole } from "@prisma/client";
+import { create } from "zustand";
+
+interface UserAccessStore {
+  userRoleInProject: UserRole;
```
Comment (dahal) [inline apps/platform/store/UserRole.ts:5]:
```suggestion
  currentRole: UserRole;
```

File: apps/platform/store/UserRole.ts [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+
+  setAccessData: (role: UserRole) => void;
```
Comment (dahal) [inline apps/platform/store/UserRole.ts:7]:
```suggestion
  setCurrentRole: (role: UserRole) => void;
```

**Review by dahal — CHANGES_REQUESTED** [review]
This is looking amazing, awesome job @ttebify , I have recommended few changes, after that its good to go 🚀

File: apps/platform/components/members/Table.tsx (hunk 1 above)
Comment (ttebify) [reply apps/platform/components/members/Table.tsx:5]: No it is not, I will remove it

File: apps/platform/components/members/Table.tsx (hunk 2 above)
Comment (ttebify) [reply apps/platform/components/members/Table.tsx:22]: Makes sense. I will do that right away

**Review by dahal — APPROVED** [review]

**Pull request merged**